Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -0,0 +1,324 @@
+instance : OrderedAddCommGroup (DivisorOn U) where
+  __ := inferInstanceAs (AddCommGroup (DivisorOn U))
+  __ := inferInstanceAs (Lattice (DivisorOn U))
+  add_le_add_left := fun _ _ _ _ ↦ by simpa [le_def]
+
+/-!
+## Restriction
+-/
+
+/-- If `V` is a subset of `U`, then a divisor on `U` restricts to a divisor in `V` by
+setting its values to zero outside of `V`. -/
+noncomputable def restrict {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) :
+    DivisorOn V where
+  toFun := by
+    classical
+    exact fun z ↦ if hz : z ∈ V then D z else 0
+  supportWithinDomain' := by
+    intro x hx
+    simp_rw [dite_eq_ite, Function.mem_support, ne_eq, ite_eq_right_iff,
+      Classical.not_imp] at hx
+    exact hx.1
+  supportDiscreteWithinDomain' := by
+    apply Filter.codiscreteWithin.mono h
+    filter_upwards [D.supportDiscreteWithinDomain]
+    intro x hx
+    simp [hx]
+
+open Classical in
+lemma restrict_apply {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) (z : 𝕜) :
+    (D.restrict h) z = if z ∈ V then D z else 0 := rfl
+
+lemma restrict_eqOn {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) :
+    Set.EqOn (D.restrict h) D V := by
+  intro _ _
+  simp_all [restrict_apply, dite_eq_ite, ite_eq_left_iff]
+
+lemma restrict_eqOn_compl {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) :
+    Set.EqOn (D.restrict h) 0 Vᶜ := by
+  intro _ hx
+  simp_all [restrict_apply, dite_eq_ite, ite_eq_left_iff, hx]
+
+/-- Restriction as a group morphism -/
+noncomputable def restrictMonoidHom {V : Set 𝕜} (h : V ⊆ U) : DivisorOn U →+ DivisorOn V where
+  toFun D := D.restrict h
+  map_zero' := by
+    ext x
+    simp [restrict_apply]
+  map_add' D₁ D₂ := by
+    ext x
+    by_cases hx : x ∈ V
+    <;> simp [restrict_apply, hx]
+
+@[simp]
+lemma restrictMonoidHom_apply {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) :
+    restrictMonoidHom h D = D.restrict h := by rfl
+
+/-- Restriction as a lattice morphism -/
+noncomputable def restrictLatticeHom {V : Set 𝕜} (h : V ⊆ U) :
+    LatticeHom (DivisorOn U) (DivisorOn V) where
+  toFun D := D.restrict h
+  map_sup' D₁ D₂ := by
+    ext x
+    by_cases hx : x ∈ V
+    <;> simp [DivisorOn.restrict_apply, hx]
+  map_inf' D₁ D₂ := by
+    ext x
+    by_cases hx : x ∈ V
+    <;> simp [DivisorOn.restrict_apply, hx]
+
+@[simp]
+lemma restrictLatticeHom_apply {V : Set 𝕜} (D : DivisorOn U) (h : V ⊆ U) :
+    restrictLatticeHom h D = D.restrict h := by rfl
+
+/-!
+## Derived invariants
+-/
+
+/-- The degree of a divisor on U is the sum of its values, or 0 if the support is infinite. -/
+noncomputable def deg (D : DivisorOn U) : ℤ := ∑ᶠ z, D z
+
+end DivisorOn
+
+/-- The counting function for a divisor -/
+noncomputable def Divisor.counting (D : Divisor 𝕜) : ℝ → ℝ :=
+  fun r ↦ ∑ᶠ z, D.restrict (by tauto : closedBall (0 : 𝕜) |r| ⊆ ⊤) z
+
+/-- The logarithmic counting function for a divisor -/
+noncomputable def Divisor.logCounting (D : Divisor 𝕜) : ℝ → ℝ :=
+  fun r ↦ ∑ᶠ z, D.restrict (by tauto : closedBall (0 : 𝕜) |r| ⊆ ⊤) z * (log r - log ‖z‖)
```

Comment:
Since there currently no theory about these definitions, I think it is better to postpone them to a later PR.

One trivial thing that I think will be good to add is `DivisorOn.single`, which is takes an `n : Int` and an `x` in `U`, and builds the divisor `n[x]`. Because once you have that, it will be easy to build arbitrary finitely-supported divisors using linear combinations.
But again, this can be done in a follow-up PR.

Reply:
Roger. I will remove this material for now and re-submit it once I get closer to formulating Nevanlinna's first main theorem. Ditto for the `DivisorOn.single` that you suggest.